File: app/src/main/java/com/lifedawn/bestweather/weathers/simplefragment/dailyforecast/SimpleDailyForecastFragment.java
package com.lifedawn.bestweather.weathers.simplefragment.dailyforecast;

import android.content.Context;
import android.os.Bundle;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentManager;

import com.lifedawn.bestweather.R;
import com.lifedawn.bestweather.commons.enums.BundleKey;
import com.lifedawn.bestweather.commons.enums.WeatherDataType;
import com.lifedawn.bestweather.commons.enums.WeatherValueType;
import com.lifedawn.bestweather.main.MyApplication;
import com.lifedawn.bestweather.weathers.FragmentType;
import com.lifedawn.bestweather.weathers.WeatherFragment;
import com.lifedawn.bestweather.weathers.comparison.dailyforecast.DailyForecastComparisonFragment;
import com.lifedawn.bestweather.weathers.detailfragment.dailyforecast.DetailDailyForecastFragment;
import com.lifedawn.bestweather.weathers.models.DailyForecastDto;
import com.lifedawn.bestweather.weathers.simplefragment.base.BaseSimpleForecastFragment;
import com.lifedawn.bestweather.weathers.view.DetailDoubleTemperatureView;
import com.lifedawn.bestweather.weathers.view.DoubleWeatherIconView;
import com.lifedawn.bestweather.weathers.view.IconTextView;
import com.lifedawn.bestweather.weathers.view.TextsView;

import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

public class SimpleDailyForecastFragment extends BaseSimpleForecastFragment {
	private static List<DailyForecastDto> dailyForecastDtoList;

	public SimpleDailyForecastFragment setDailyForecastDtoList(List<DailyForecastDto> dailyForecastDtoList) {
		this.dailyForecastDtoList = dailyForecastDtoList;
		return this;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		needCompare = true;
	}

	@Override
	public void onInflateFinished(@NonNull View view, int resid, @Nullable ViewGroup parent) {
		super.onInflateFinished(view, resid, parent);

		binding.weatherCardViewHeader.forecastName.setText(R.string.daily_forecast);
		binding.weatherCardViewHeader.compareForecast.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				if (availableNetwork()) {
					DailyForecastComparisonFragment comparisonFragment = new DailyForecastComparisonFragment();
					comparisonFragment.setArguments(bundle);

					String tag = getString(R.string.tag_comparison_fragment);
					FragmentManager fragmentManager = getParentFragment().getParentFragmentManager();

					fragmentManager.beginTransaction().hide(
							fragmentManager.findFragmentByTag(WeatherFragment.class.getName())).add(R.id.fragment_container,
							comparisonFragment, tag).addToBackStack(tag).commit();
				}
			}
		});

		binding.weatherCardViewHeader.detailForecast.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				Bundle arguments = new Bundle();
				arguments.putString(BundleKey.AddressName.name(), bundle.getString(BundleKey.AddressName.name()));
				arguments.putSerializable(BundleKey.TimeZone.name(), bundle.getSerializable(BundleKey.TimeZone.name()));
				arguments.putSerializable(BundleKey.WeatherProvider.name(), mainWeatherProviderType);

				DetailDailyForecastFragment detailDailyForecastFragment = new DetailDailyForecastFragment();
				DetailDailyForecastFragment.setDailyForecastDtoList(dailyForecastDtoList);
				detailDailyForecastFragment.setArguments(arguments);

				String tag = getString(R.string.tag_detail_daily_forecast_fragment);
				FragmentManager fragmentManager = getParentFragment().getParentFragmentManager();

				fragmentManager.beginTransaction().hide(
						fragmentManager.findFragmentByTag(WeatherFragment.class.getName())).add(R.id.fragment_container,
						detailDailyForecastFragment, tag).addToBackStack(tag).commit();
			}
		});

		setValuesToViews();
	}

	@Override
	public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
	}

	@Override
	public void setValuesToViews() {
		super.setValuesToViews();
		// 날짜 ,낮과 밤의 날씨상태, 강수확률, 강우량, 강설량, 최저/최고 기온

		MyApplication.getExecutorService().submit(() -> {
			Context context = getContext();

			final int WEATHER_ROW_HEIGHT = (int) context.getResources().getDimension(R.dimen.singleWeatherIconValueRowHeightInSC);
			final int TEMP_ROW_HEIGHT = (int) context.getResources().getDimension(R.dimen.doubleTemperatureRowHeightInSC);

			final int COLUMN_COUNT = dailyForecastDtoList.size();
			final int COLUMN_WIDTH = (int) context.getResources().getDimension(R.dimen.valueColumnWidthInSDailyOwm);
			final int VIEW_WIDTH = COLUMN_COUNT * COLUMN_WIDTH;

			DoubleWeatherIconView weatherIconRow = new DoubleWeatherIconView(context, FragmentType.Simple, VIEW_WIDTH, WEATHER_ROW_HEIGHT,
					COLUMN_WIDTH);
			IconTextView probabilityOfPrecipitationRow = new IconTextView(context, FragmentType.Simple, VIEW_WIDTH,
					COLUMN_WIDTH, R.drawable.pop);
			IconTextView rainVolumeRow = new IconTextView(context, FragmentType.Simple, VIEW_WIDTH,
					COLUMN_WIDTH, R.drawable.raindrop);
			IconTextView snowVolumeRow = new IconTextView(context, FragmentType.Simple, VIEW_WIDTH,
					COLUMN_WIDTH, R.drawable.snowparticle);

			//시각 --------------------------------------------------------------------------
			List<String> dateList = new ArrayList<>();
			DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("M.d\nE");
			//날씨 아이콘
			List<DoubleWeatherIconView.WeatherIconObj> weatherIconObjList = new ArrayList<>();
			//기온, 강수확률, 강수량
			List<Integer> minTempList = new ArrayList<>();
			List<Integer> maxTempList = new ArrayList<>();
			List<String> popList = new ArrayList<>();
			List<String> rainVolumeList = new ArrayList<>();
			List<String> snowVolumeList = new ArrayList<>();

			final String tempDegree = MyApplication.VALUE_UNIT_OBJ.getTempUnitText();
			final String mm = "mm";
			final String cm = "cm";

			boolean haveSnow = false;
			boolean haveRain = false;

			float rainVolume = 0f;
			float snowVolume = 0f;

			final String zeroPrecipitationFormat = "%.1f";

			for (DailyForecastDto item : dailyForecastDtoList) {
				if (!item.isAvailable_toMakeMinMaxTemp())
					break;

				rainVolume = 0f;
				snowVolume = 0f;

				dateList.add(item.getDate().format(dateTimeFormatter));
				minTempList.add(Integer.parseInt(item.getMinTemp().replace(tempDegree, "")));
				maxTempList.add(Integer.parseInt(item.getMaxTemp().replace(tempDegree, "")));

				if (item.getValuesList().size() == 1) {
					popList.add(item.getValuesList().get(0).getPop());
					weatherIconObjList.add(new DoubleWeatherIconView.WeatherIconObj(ContextCompat.getDrawable(context,
							item.getValuesList().get(0).getWeatherIcon()), item.getValuesList().get(0).getWeatherDescription()));

					if (item.getValuesList().get(0).getRainVolume() != null) {
						rainVolume = Float.parseFloat(item.getValuesList().get(0).getRainVolume().replace(mm, "").replace(cm, ""));
					}
					if (item.getValuesList().get(0).getSnowVolume() != null) {
						snowVolume = Float.parseFloat(item.getValuesList().get(0).getSnowVolume().replace(mm, "").replace(cm, ""));
					}
				} else if (item.getValuesList().size() == 2) {
					popList.add(item.getValuesList().get(0).getPop() + "/" + item.getValuesList().get(1).getPop());
					weatherIconObjList.add(new DoubleWeatherIconView.WeatherIconObj(ContextCompat.getDrawable(context,
							item.getValuesList().get(0).getWeatherIcon()),
							ContextCompat.getDrawable(context, item.getValuesList().get(1).getWeatherIcon()),
							item.getValuesList().get(0).getWeatherDescription(),
							item.getValuesList().get(1).getWeatherDescription()));

					if (item.getValuesList().get(0).getRainVolume() != null || item.getValuesList().get(1).getRainVolume() != null) {
						rainVolume = Float.parseFloat(item.getValuesList().get(0).getRainVolume().replace(mm, "").replace(cm, ""))
								+ Float.parseFloat(item.getValuesList().get(1).getRainVolume().replace(mm, "").replace(cm, ""));
					}
					if (item.getValuesList().get(0).getSnowVolume() != null || item.getValuesList().get(1).getSnowVolume() != null) {
						snowVolume = Float.parseFloat(item.getValuesList().get(0).getSnowVolume().replace(mm, "").replace(cm, ""))
								+ Float.parseFloat(item.getValuesList().get(1).getSnowVolume().replace(mm, "").replace(cm, ""));
					}
				} else if (item.getValuesList().size() == 4) {
					weatherIconObjList.add(new DoubleWeatherIconView.WeatherIconObj(ContextCompat.getDrawable(context,
							item.getValuesList().get(1).getWeatherIcon()),
							ContextCompat.getDrawable(context, item.getValuesList().get(2).getWeatherIcon()),
							item.getValuesList().get(1).getWeatherDescription(),
							item.getValuesList().get(2).getWeatherDescription()));

					if (item.getValuesList().get(0).isHasPrecipitationVolume() || item.getValuesList().get(1).isHasPrecipitationVolume() ||
							item.getValuesList().get(2).isHasPrecipitationVolume() || item.getValuesList().get(3).isHasPrecipitationVolume()) {
						rainVolume = Float.parseFloat(item.getValuesList().get(0).getPrecipitationVolume().replace(mm, "").replace(cm, ""))
								+ Float.parseFloat(item.getValuesList().get(1).getPrecipitationVolume().replace(mm, "").replace(cm, ""))
								+ Float.parseFloat(item.getValuesList().get(2).getPrecipitationVolume().replace(mm, "").replace(cm, ""))
								+ Float.parseFloat(item.getValuesList().get(3).getPrecipitationVolume().replace(mm, "").replace(cm, ""));
					}

				}

				rainVolumeList.add(String.format(Locale.getDefault(), zeroPrecipitationFormat, rainVolume));
				snowVolumeList.add(String.format(Locale.getDefault(), zeroPrecipitationFormat, snowVolume));

				if (!haveRain) {
					if (rainVolume > 0f) {
						haveRain = true;
					}
				}
				if (!haveSnow) {
					if (snowVolume > 0f) {
						haveSnow = true;
					}
				}
			}

			weatherIconRow.setIcons(weatherIconObjList);
			probabilityOfPrecipitationRow.setValueList(popList);
			rainVolumeRow.setValueList(rainVolumeList);
			snowVolumeRow.setValueList(snowVolumeList);

			TextsView dateRow = new TextsView(context, VIEW_WIDTH, COLUMN_WIDTH, dateList);
			DetailDoubleTemperatureView tempRow = new DetailDoubleTemperatureView(getContext(), FragmentType.Simple, VIEW_WIDTH,
					TEMP_ROW_HEIGHT, COLUMN_WIDTH, minTempList, maxTempList);

			LinearLayout.LayoutParams rowLayoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
					ViewGroup.LayoutParams.WRAP_CONTENT);

			LinearLayout.LayoutParams dateRowLayoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
					ViewGroup.LayoutParams.WRAP_CONTENT);
			dateRowLayoutParams.gravity = Gravity.CENTER_VERTICAL;
			dateRowLayoutParams.bottomMargin = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 4f, getResources().getDisplayMetrics());

			if (textSizeMap.containsKey(WeatherValueType.date)) {
				dateRow.setValueTextSize(textSizeMap.get(WeatherValueType.date));
			}
			if (textSizeMap.containsKey(WeatherValueType.pop)) {
				probabilityOfPrecipitationRow.setValueTextSize(textSizeMap.get(WeatherValueType.pop));
			}
			if (textSizeMap.containsKey(WeatherValueType.rainVolume)) {
				rainVolumeRow.setValueTextSize(textSizeMap.get(WeatherValueType.rainVolume));
			}
			if (textSizeMap.containsKey(WeatherValueType.snowVolume)) {
				snowVolumeRow.setValueTextSize(textSizeMap.get(WeatherValueType.snowVolume));
			}
			if (textSizeMap.containsKey(WeatherValueType.temp)) {
				tempRow.setTempTextSize(textSizeMap.get(WeatherValueType.temp));
			}

			if (textColorMap.containsKey(WeatherValueType.date)) {
				dateRow.setValueTextColor(textColorMap.get(WeatherValueType.date));
			}
			if (textColorMap.containsKey(WeatherValueType.pop)) {
				probabilityOfPrecipitationRow.setTextColor(textColorMap.get(WeatherValueType.pop));
			}
			if (textColorMap.containsKey(WeatherValueType.rainVolume)) {
				rainVolumeRow.setTextColor(textColorMap.get(WeatherValueType.rainVolume));
			}
			if (textColorMap.containsKey(WeatherValueType.snowVolume)) {
				snowVolumeRow.setTextColor(textColorMap.get(WeatherValueType.snowVolume));
			}
			if (textColorMap.containsKey(WeatherValueType.temp)) {
				tempRow.setTextColor(textColorMap.get(WeatherValueType.temp));
			}

			boolean finalHaveRain = haveRain;
			boolean finalHaveSnow = haveSnow;
			Objects.requireNonNull(requireActivity()).runOnUiThread(() -> {
				binding.forecastView.addView(dateRow, dateRowLayoutParams);
				binding.forecastView.addView(weatherIconRow, rowLayoutParams);
				binding.forecastView.addView(probabilityOfPrecipitationRow, rowLayoutParams);
				if (finalHaveRain) {
					binding.forecastView.addView(rainVolumeRow, rowLayoutParams);
				}
				if (finalHaveSnow) {
					binding.forecastView.addView(snowVolumeRow, rowLayoutParams);
				}

				LinearLayout.LayoutParams tempRowLayoutParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
						ViewGroup.LayoutParams.WRAP_CONTENT);
				tempRowLayoutParams.topMargin = (int) getResources().getDimension(R.dimen.tempTopMargin);
				binding.forecastView.addView(tempRow, tempRowLayoutParams);

				createValueUnitsDescription(mainWeatherProviderType, finalHaveRain, finalHaveSnow);

				onFinishedSetData();
			});
		});


	}

}

File: app/src/main/java/com/lifedawn/bestweather/ui/weathers/simplefragment/dailyforecast/SimpleDailyForecastFragment.kt
package com.lifedawn.bestweather.ui.weathers.simplefragment.dailyforecast

import android.os.Bundle
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.core.content.ContextCompat
import com.lifedawn.bestweather.R
import com.lifedawn.bestweather.commons.constants.BundleKey
import com.lifedawn.bestweather.commons.constants.WeatherValueType
import com.lifedawn.bestweather.data.MyApplication
import com.lifedawn.bestweather.data.local.weather.models.DailyForecastDto
import com.lifedawn.bestweather.ui.weathers.FragmentType
import com.lifedawn.bestweather.ui.weathers.WeatherFragment
import com.lifedawn.bestweather.ui.weathers.comparison.dailyforecast.DailyForecastComparisonFragment
import com.lifedawn.bestweather.ui.weathers.detailfragment.base.BaseDetailDailyForecastFragment
import com.lifedawn.bestweather.ui.weathers.detailfragment.dailyforecast.DetailDailyForecastFragment
import com.lifedawn.bestweather.ui.weathers.simplefragment.base.BaseSimpleForecastFragment
import com.lifedawn.bestweather.ui.weathers.customview.DetailDoubleTemperatureView
import com.lifedawn.bestweather.ui.weathers.customview.DoubleWeatherIconView
import com.lifedawn.bestweather.ui.weathers.customview.IconTextView
import com.lifedawn.bestweather.ui.weathers.customview.TextsView
import java.time.format.DateTimeFormatter
import java.util.*

class SimpleDailyForecastFragment : BaseSimpleForecastFragment() {
    fun setDailyForecastDtoList(dailyForecastDtoList: List<DailyForecastDto>?): SimpleDailyForecastFragment {
        Companion.dailyForecastDtoList = dailyForecastDtoList
        return this
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        needCompare = true
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onInflateFinished(view: View, resid: Int, parent: ViewGroup?) {
        super.onInflateFinished(view, resid, parent)
        binding!!.weatherCardViewHeader.forecastName.setText(R.string.daily_forecast)
        binding!!.weatherCardViewHeader.compareForecast.setOnClickListener {
            if (availableNetwork()) {
                val comparisonFragment = DailyForecastComparisonFragment()
                comparisonFragment.arguments = bundle
                val tag = getString(R.string.tag_comparison_fragment)
                val fragmentManager = parentFragment!!.parentFragmentManager
                fragmentManager.beginTransaction().hide(
                    fragmentManager.findFragmentByTag(WeatherFragment::class.java.name)!!
                ).add(
                    R.id.fragment_container,
                    comparisonFragment, tag
                ).addToBackStack(tag).commit()
            }
        }
        binding!!.weatherCardViewHeader.detailForecast.setOnClickListener {
            val arguments = Bundle()
            arguments.putString(BundleKey.AddressName.name, bundle!!.getString(BundleKey.AddressName.name))
            arguments.putSerializable(BundleKey.TimeZone.name, bundle!!.getSerializable(BundleKey.TimeZone.name))
            arguments.putSerializable(BundleKey.WeatherProvider.name, mainWeatherProviderType)
            val copiedList: List<DailyForecastDto> = ArrayList(dailyForecastDtoList)
            Collections.copy(copiedList, dailyForecastDtoList)
            val detailDailyForecastFragment = DetailDailyForecastFragment()
            BaseDetailDailyForecastFragment.setDailyForecastDtoList(copiedList)
            detailDailyForecastFragment.arguments = arguments
            val tag = getString(R.string.tag_detail_daily_forecast_fragment)
            val fragmentManager = parentFragment!!.parentFragmentManager
            fragmentManager.beginTransaction().hide(
                fragmentManager.findFragmentByTag(WeatherFragment::class.java.name)!!
            ).add(
                R.id.fragment_container,
                detailDailyForecastFragment, tag
            ).addToBackStack(tag).commit()
        }
        setValuesToViews()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
    }

    override fun setValuesToViews() {
        super.setValuesToViews()
        // 날짜 ,낮과 밤의 날씨상태, 강수확률, 강우량, 강설량, 최저/최고 기온
        MyApplication.getExecutorService().submit {
            val context = requireContext().applicationContext
            val WEATHER_ROW_HEIGHT = context.resources.getDimension(R.dimen.singleWeatherIconValueRowHeightInSC).toInt()
            val TEMP_ROW_HEIGHT = context.resources.getDimension(R.dimen.doubleTemperatureRowHeightInSC).toInt()
            val COLUMN_COUNT = dailyForecastDtoList!!.size
            val COLUMN_WIDTH = context.resources.getDimension(R.dimen.valueColumnWidthInSDailyOwm).toInt()
            val VIEW_WIDTH = COLUMN_COUNT * COLUMN_WIDTH
            val weatherIconRow = DoubleWeatherIconView(
                context, FragmentType.Simple, VIEW_WIDTH, WEATHER_ROW_HEIGHT,
                COLUMN_WIDTH
            )
            val probabilityOfPrecipitationRow = IconTextView(
                context, FragmentType.Simple, VIEW_WIDTH,
                COLUMN_WIDTH, R.drawable.pop
            )
            val rainVolumeRow = IconTextView(
                context, FragmentType.Simple, VIEW_WIDTH,
                COLUMN_WIDTH, R.drawable.raindrop
            )
            val snowVolumeRow = IconTextView(
                context, FragmentType.Simple, VIEW_WIDTH,
                COLUMN_WIDTH, R.drawable.snowparticle
            )
            customViewList.add(weatherIconRow)
            customViewList.add(probabilityOfPrecipitationRow)
            customViewList.add(rainVolumeRow)
            customViewList.add(snowVolumeRow)

            //시각 --------------------------------------------------------------------------
            val dateList: MutableList<String> = ArrayList()
            val dateTimeFormatter = DateTimeFormatter.ofPattern("M.d\nE")
            //날씨 아이콘
            val weatherIconObjList: MutableList<DoubleWeatherIconView.WeatherIconObj> = ArrayList()
            //기온, 강수확률, 강수량
            val minTempList: MutableList<Int> = ArrayList()
            val maxTempList: MutableList<Int> = ArrayList()
            val popList: MutableList<String> = ArrayList()
            val rainVolumeList: MutableList<String> = ArrayList()
            val snowVolumeList: MutableList<String> = ArrayList()
            val tempDegree = MyApplication.VALUE_UNIT_OBJ.tempUnitText
            val mm = "mm"
            val cm = "cm"
            var haveSnow = false
            var haveRain = false
            var rainVolume = 0f
            var snowVolume = 0f
            val zeroPrecipitationFormat = "%.1f"
            for ((date, valuesList, minTemp, maxTemp, _, _, isAvailable_toMakeMinMaxTemp) in dailyForecastDtoList!!) {
                if (!isAvailable_toMakeMinMaxTemp) break
                rainVolume = 0f
                snowVolume = 0f
                dateList.add(date.format(dateTimeFormatter))
                minTempList.add(minTemp.replace(tempDegree, "").toInt())
                maxTempList.add(maxTemp.replace(tempDegree, "").toInt())
                if (valuesList.size == 1) {
                    popList.add(valuesList[0].pop)
                    weatherIconObjList.add(
                        DoubleWeatherIconView.WeatherIconObj(
                            ContextCompat.getDrawable(
                                context,
                                valuesList[0].weatherIcon
                            ), valuesList[0].weatherDescription
                        )
                    )
                    if (valuesList[0].rainVolume != null) {
                        rainVolume = valuesList[0].rainVolume.replace(mm, "").replace(cm, "").toFloat()
                    }
                    if (valuesList[0].snowVolume != null) {
                        snowVolume = valuesList[0].snowVolume.replace(mm, "").replace(cm, "").toFloat()
                    }
                } else if (valuesList.size == 2) {
                    popList.add(valuesList[0].pop + "/" + valuesList[1].pop)
                    weatherIconObjList.add(
                        DoubleWeatherIconView.WeatherIconObj(
                            ContextCompat.getDrawable(
                                context,
                                valuesList[0].weatherIcon
                            ),
                            ContextCompat.getDrawable(context, valuesList[1].weatherIcon),
                            valuesList[0].weatherDescription,
                            valuesList[1].weatherDescription
                        )
                    )
                    if (valuesList[0].rainVolume != null || valuesList[1].rainVolume != null) {
                        rainVolume =
                            valuesList[0].rainVolume.replace(mm, "").replace(cm, "").toFloat() + valuesList[1].rainVolume.replace(mm, "")
                                .replace(cm, "").toFloat()
                    }
                    if (valuesList[0].snowVolume != null || valuesList[1].snowVolume != null) {
                        snowVolume =
                            valuesList[0].snowVolume.replace(mm, "").replace(cm, "").toFloat() + valuesList[1].snowVolume.replace(mm, "")
                                .replace(cm, "").toFloat()
                    }
                } else if (valuesList.size == 4) {
                    weatherIconObjList.add(
                        DoubleWeatherIconView.WeatherIconObj(
                            ContextCompat.getDrawable(
                                context,
                                valuesList[1].weatherIcon
                            ),
                            ContextCompat.getDrawable(context, valuesList[2].weatherIcon),
                            valuesList[1].weatherDescription,
                            valuesList[2].weatherDescription
                        )
                    )
                    if (valuesList[0].isHasPrecipitationVolume || valuesList[1].isHasPrecipitationVolume ||
                        valuesList[2].isHasPrecipitationVolume || valuesList[3].isHasPrecipitationVolume
                    ) {
                        rainVolume = valuesList[0].precipitationVolume.replace(mm, "").replace(cm, "")
                            .toFloat() + valuesList[1].precipitationVolume.replace(mm, "").replace(cm, "")
                            .toFloat() + valuesList[2].precipitationVolume.replace(mm, "").replace(cm, "")
                            .toFloat() + valuesList[3].precipitationVolume.replace(mm, "").replace(cm, "").toFloat()
                    }
                }
                rainVolumeList.add(String.format(Locale.getDefault(), zeroPrecipitationFormat, rainVolume))
                snowVolumeList.add(String.format(Locale.getDefault(), zeroPrecipitationFormat, snowVolume))
                if (!haveRain) {
                    if (rainVolume > 0f) {
                        haveRain = true
                    }
                }
                if (!haveSnow) {
                    if (snowVolume > 0f) {
                        haveSnow = true
                    }
                }
            }
            weatherIconRow.setIcons(weatherIconObjList)
            probabilityOfPrecipitationRow.valueList = popList
            rainVolumeRow.valueList = rainVolumeList
            snowVolumeRow.valueList = snowVolumeList
            val dateRow = TextsView(context, VIEW_WIDTH, COLUMN_WIDTH, dateList)
            val tempRow = DetailDoubleTemperatureView(
                requireContext().applicationContext, FragmentType.Simple, VIEW_WIDTH,
                TEMP_ROW_HEIGHT, COLUMN_WIDTH, minTempList, maxTempList
            )
            customViewList.add(dateRow)
            customViewList.add(tempRow)
            val rowLayoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            val dateRowLayoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            dateRowLayoutParams.gravity = Gravity.CENTER_VERTICAL
            dateRowLayoutParams.bottomMargin = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 4f, resources.displayMetrics).toInt()
            if (textSizeMap!!.containsKey(WeatherValueType.date)) {
                dateRow.setValueTextSize(textSizeMap!![WeatherValueType.date]!!)
            }
            if (textSizeMap!!.containsKey(WeatherValueType.pop)) {
                probabilityOfPrecipitationRow.setValueTextSize(textSizeMap!![WeatherValueType.pop]!!)
            }
            if (textSizeMap!!.containsKey(WeatherValueType.rainVolume)) {
                rainVolumeRow.setValueTextSize(textSizeMap!![WeatherValueType.rainVolume]!!)
            }
            if (textSizeMap!!.containsKey(WeatherValueType.snowVolume)) {
                snowVolumeRow.setValueTextSize(textSizeMap!![WeatherValueType.snowVolume]!!)
            }
            if (textSizeMap!!.containsKey(WeatherValueType.temp)) {
                tempRow.setTempTextSize(textSizeMap!![WeatherValueType.temp]!!)
            }
            if (textColorMap!!.containsKey(WeatherValueType.date)) {
                dateRow.setValueTextColor(textColorMap!![WeatherValueType.date]!!)
            }
            if (textColorMap!!.containsKey(WeatherValueType.pop)) {
                probabilityOfPrecipitationRow.setTextColor(textColorMap!![WeatherValueType.pop]!!)
            }
            if (textColorMap!!.containsKey(WeatherValueType.rainVolume)) {
                rainVolumeRow.setTextColor(textColorMap!![WeatherValueType.rainVolume]!!)
            }
            if (textColorMap!!.containsKey(WeatherValueType.snowVolume)) {
                snowVolumeRow.setTextColor(textColorMap!![WeatherValueType.snowVolume]!!)
            }
            if (textColorMap!!.containsKey(WeatherValueType.temp)) {
                tempRow.setTextColor(textColorMap!![WeatherValueType.temp]!!)
            }
            val finalHaveRain = haveRain
            val finalHaveSnow = haveSnow
            try {
                requireActivity().runOnUiThread {
                    binding!!.forecastView.addView(dateRow, dateRowLayoutParams)
                    binding!!.forecastView.addView(weatherIconRow, rowLayoutParams)
                    binding!!.forecastView.addView(probabilityOfPrecipitationRow, rowLayoutParams)
                    if (finalHaveRain) {
                        binding!!.forecastView.addView(rainVolumeRow, rowLayoutParams)
                    }
                    if (finalHaveSnow) {
                        binding!!.forecastView.addView(snowVolumeRow, rowLayoutParams)
                    }
                    val tempRowLayoutParams = LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    )
                    tempRowLayoutParams.topMargin = resources.getDimension(R.dimen.tempTopMargin).toInt()
                    binding!!.forecastView.addView(tempRow, tempRowLayoutParams)
                    createValueUnitsDescription(mainWeatherProviderType!!, finalHaveRain, finalHaveSnow)
                    onFinishedSetData()
                }
            } catch (e: Exception) {
            }
        }
    }

    companion object {
        private var dailyForecastDtoList: List<DailyForecastDto>? = null
    }
}